### PR TITLE
Add Network test service and Vpn android app

### PR DESCRIPTION
**Pull request by DexterDreeeam**
This PR is for Android Device Network test by Android VPN Service.

- Add VPN service to Hydra Lab Client app
- Add Network Test parameters in Test Task Spec entity
- Change Espresso Flow to enable Network test

image

## Pull Request Checklist

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code compiles correctly with all tests are passed.
- [x] I've read the [contributing guide](https://github.com/microsoft/HydraLab/blob/main/CONTRIBUTING.md#making-changes-to-the-code) and followed the recommended practices.
- [ ] [Wikis](https://github.com/microsoft/HydraLab/wiki) or [README](https://github.com/microsoft/HydraLab/blob/main/README.md) have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?
*If this introduces a breaking change for Hydra Lab users, please describe the impact and migration path.*

- [ ] Yes
- [ ] No

## Pull Request Description

**A few words to explain your changes**:  

### How you tested it
*Please make sure the change is tested, you can test it by adding UTs, do local test and share the screenshots, etc.*

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Technical design
- [ ] Build related changes
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Code style update (formatting, renaming) or Documentation content changes
- [ ] Other (please describe): 

### Feature UI screenshots or Technical design diagrams
*If this is a relatively large or complex change, kick it off by drawing the tech design with PlantUML and explaining why you chose the solution you did and what alternatives you considered, etc...*

### CODE REVIEW

File: common/src/main/java/com/microsoft/hydralab/common/network/DummyNetworkMonitor.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import java.io.File;
+
+public class DummyNetworkMonitor implements NetworkMonitor {
+
+    public DummyNetworkMonitor() {
+    }
+
+    @Override
+    public void start() {
+    }
```
Comment (zhou9584) [inline common/src/main/java/com/microsoft/hydralab/common/network/DummyNetworkMonitor.java:17]: Add some logs to declare this class is a dummy monitor.

File: common/src/main/java/com/microsoft/hydralab/common/network/AndroidNetworkMonitor.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    public void stop() {
+        // stop vpn
+        String command_stop = "adb shell am start";
+        command_stop += " -a com.microsoft.hydralab.android.client.vpn.STOP";
+        command_stop += " -n com.microsoft.hydralab.android.client/.MainActivity";
+        ShellUtils.execLocalCommandWithResult(command_stop, logger);
+        ThreadUtils.safeSleep(2000);
+
+        // pull result
+        String local_dump_path = resultFolder.getAbsolutePath() + DumpPath;
```
Comment (zhou9584) [inline common/src/main/java/com/microsoft/hydralab/common/network/AndroidNetworkMonitor.java:87]: You can reuse com.microsoft.hydralab.common.util.ADBOperateUtil#pullFileToDir to pull file, there is a retry mechanism in this method.

File: common/src/main/java/com/microsoft/hydralab/common/network/DummyNetworkMonitor.java (hunk 1 above)
Comment (DexterDreeeam) [reply common/src/main/java/com/microsoft/hydralab/common/network/DummyNetworkMonitor.java:21]: fixed

File: common/src/main/java/com/microsoft/hydralab/common/network/AndroidNetworkMonitor.java (hunk 2 above)
Comment (DexterDreeeam) [reply common/src/main/java/com/microsoft/hydralab/common/network/AndroidNetworkMonitor.java]: fixed

**Review by olivershen-wow — APPROVED** [review]
LGTM

File: android_client/app/build.gradle [hunk 3]
```diff
@@ -37,6 +50,9 @@ android {
 }
 
 dependencies {
+    implementation 'androidx.core:core-ktx:1.10.1'
+    implementation 'androidx.lifecycle:lifecycle-runtime-ktx:2.6.1'
+    implementation 'androidx.annotation:annotation-jvm:+'
```
Comment (DexterDreeeam) [inline android_client/app/build.gradle:55]: use fixed version instead dynamic version ---- Jian

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/espresso/EspressoTestInfoProcessorListener.java [hunk 4]
```diff
@@ -71,6 +74,9 @@ public void startRecording(int maxTime) {
         final String initializing = "Initializing";
         testRunDeviceOrchestrator.setRunningTestName(testRunDevice, initializing);
         testRun.addNewTimeTag(initializing, 0);
+        if (testTask.isEnableNetworkMonitor()) {
```
Comment (taoran6) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/espresso/EspressoTestInfoProcessorListener.java:77]: In the future we are going to disable the screen recording for long-time test run. Going forward, the logics of starting VPN should be removed from `startRecording()`. cc @zhou9584

**Review by taoran6 — APPROVED** [review]
Others LGTM.

File: android_client/app/src/main/java/com/microsoft/hydralab/android/client/MainActivity.java [hunk 5]
```diff
@@ -139,6 +146,27 @@ protected void onCreate(Bundle savedInstanceState) {
                         .getBoolean(getResources().getResourceEntryName(mAudioSwitch.getId()), false));
     }
 
+    @Override
+    protected void onNewIntent(Intent intent) {
```
Comment (DexterDreeeam) [inline android_client/app/src/main/java/com/microsoft/hydralab/android/client/MainActivity.java:150]: move to onResume and remove SingleTop

**Review by TedaLIEz — DISMISSED** [review]
Approved with android_client

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/espresso/EspressoTestInfoProcessorListener.java (hunk 4 above)
Comment (DexterDreeeam) [reply agent/src/main/java/com/microsoft/hydralab/agent/runner/espresso/EspressoTestInfoProcessorListener.java:77]: yes

**Review by zhou9584 — APPROVED** [review]

**Review by lifesaver0129 — APPROVED** [review]

**Pull request merged**